src/Tasks/StateFileBase.cs

```diff
@@ -77,32 +78,35 @@ internal static StateFileBase DeserializeCache(string stateFile, TaskLoggingHelp
-                            // If there is an invalid cast, a message rather than a warning should be emitted.
-                            log.LogMessageFromResources("General.CouldNotReadStateFileMessage", stateFile, log.FormatResourceString("General.IncompatibleStateFileType"));
-                        }
-
-                        if ((retVal != null) && (!requiredReturnType.IsInstanceOfType(retVal)))
+                        deserializedObject = formatter.Deserialize(s);
+                        retVal = deserializedObject as T;
+                    }
+                    // If the deserialized object is null then there would be no cast error but retVal would still be null
+                    // only log the message if there would have been a cast error
+                    if (retVal == null && deserializedObject != null)
+                    {
+                        // When upgrading to Visual Studio 2008 and running the build for the first time the resource cache files are replaced which causes a cast error due
+                        // to a new version number on the tasks class. "Unable to cast object of type 'Microsoft.Build.Tasks.SystemState' to type 'Microsoft.Build.Tasks.StateFileBase".
+                        // If there is an invalid cast, a message rather than a warning should be emitted.
+                        log.LogMessageFromResources("General.CouldNotReadStateFileMessage", stateFile, log.FormatResourceString("General.IncompatibleStateFileType"));
+                    }
+                    else if (retVal != null && !(retVal is T))
+                    {
+                        if (logWarnings)
                         {
-                            log.LogWarningWithCodeFromResources("General.CouldNotReadStateFile", stateFile,
-                                log.FormatResourceString("General.IncompatibleStateFileType"));
-                            retVal = null;
+                            log.LogWarningWithCodeFromResources("General.CouldNotReadStateFile", stateFile, log.FormatResourceString("General.IncompatibleStateFileType"));
                         }
-
-                        // If we get back a valid object and internals were changed, things are likely to be null. Check the version before we use it.
-                        if (retVal != null && retVal._serializedVersion != CurrentSerializationVersion)
+                        else
                         {
-                            log.LogMessageFromResources("General.CouldNotReadStateFileMessage", stateFile, log.FormatResourceString("General.IncompatibleStateFileType"));
-                            retVal = null;
+                            log.LogMessageFromResources("General.CouldNotReadStateFile", stateFile, log.FormatResourceString("General.IncompatibleStateFileType"));
```

Comment:
I don't see what you mean? It looks to me like the retVal._serializedVersion != CurrentSerializationVersion check is after logging a warning (always) in the old code. The changes here are:
1) The conditions were mutually exclusive --> switch to else if
2) Make function parameterized instead of explicitly taking a type as an argument
3) Give the option of not logging warnings